src/engraving/playback/playbackmodel.cpp

```diff
@@ -258,52 +258,74 @@ void PlaybackModel::updateEvents(const int tickFrom, const int tickTo, const tra
                         continue;
                     }
 
-                    Ms::ID partId = item->part()->id();
+                    int segmentStartTick = segment->tick().ticks();
+                    int segmentEndTick = segmentStartTick + segment->ticks().ticks();
 
-                    if (changedPartIdSet.find(partId) == changedPartIdSet.cend()) {
+                    if (segmentStartTick > tickTo || segmentEndTick <= tickFrom) {
                         continue;
                     }
 
-                    InstrumentTrackId trackId = idKey(item);
+                    for (int voice = 0; voice < Ms::VOICES; ++voice) {
+                        int track = staff->idx() * Ms::VOICES + voice;
+                        const Ms::EngravingItem* item = segment->element(track);
 
-                    if (!trackId.isValid()) {
-                        continue;
-                    }
+                        if (!item || !item->isChordRest() || !item->part()) {
+                            continue;
+                        }
 
-                    const PlaybackContext& ctx = m_playbackCtxMap[trackId];
+                        InstrumentTrackId trackId = idKey(item);
 
-                    ArticulationsProfilePtr profile = profilesRepository()->defaultProfile(m_playbackDataMap[trackId].setupData.category);
-                    if (!profile) {
-                        LOGE() << "unsupported instrument family: " << partId;
-                        continue;
+                        if (!trackId.isValid()) {
+                            continue;
+                        }
+
+                        const PlaybackContext& ctx = m_playbackCtxMap[trackId];
+
+                        ArticulationsProfilePtr profile
+                            = profilesRepository()->defaultProfile(m_playbackDataMap[trackId].setupData.category);
+                        if (!profile) {
+                            LOGE() << "unsupported instrument family: " << partId;
+                            continue;
+                        }
+
+                        m_renderer.render(item, measureTicksOffset, ctx.appliableDynamicLevel(segmentStartTick + measureTicksOffset),
+                                          ctx.persistentArticulationType(segmentStartTick + measureTicksOffset), std::move(profile),
+                                          m_playbackDataMap[trackId].originEvents);
+
+                        collectChangesTracks(trackId, trackChanges);
                     }
+                }
+            }
 
-                    m_renderer.render(item, tickPositionOffset, ctx.appliableDynamicLevel(segmentStartTick + tickPositionOffset),
-                                      ctx.persistentArticulationType(segmentStartTick + tickPositionOffset), std::move(profile),
-                                      m_playbackDataMap[trackId].originEvents);
+            for (Ms::Segment* segment = measure->first(); segment; segment = segment->next()) {
```

Comment:
Don't quite understand the motivation. Now we have 2 loops for segments for each repeatSegment.
Performance is highly critical here, so I'd propose you to return previous implementation.

Reply:
Let's give ourselves an overview:

Old situation:
```
for each measure:
    for each segment:
        for each track (segment->elist())
            render the element in that segment on that track
        render the metronome for that segment
```

New situation: 
```
for each measure: 
    for each staff:
        determine the "playback measure" for the combination of the measure and the staff
        for each segment of the "playback measure":
            for each track of that staff:
                render the element in that segment on that track
    
    (not specific to any staff)
    for each segment of the measure:
        render the metronome for that segment
```
The motivation is as follows: for each measure, we need to handle each staff separately, because we need to determine the playback measure based on the combination of staff and measure. And we can't do anything with segments before we have determined the playback measure. So the segments loop really needs to be nested inside the staves loop. And then we need one more segments loop to render the metronome, which is not specific to any staff.